src/lib/ValidatorCountsReport.sol

```diff
@@ -6,20 +6,33 @@ pragma solidity 0.8.24;
 library ValidatorCountsReport {
     error InvalidReportData();
 
-    // TODO: consider joining with validate
-    function countOperators(
-        bytes calldata ids
-    ) internal pure returns (uint256) {
-        return ids.length / 8;
-    }
+    function safeCountOperators(
+        bytes calldata ids,
+        bytes calldata counts
+    ) internal pure returns (uint256 count) {
+        assembly ("memory-safe") {
+            // counts.length / 16 != ids.length / 8
+            if iszero(eq(div(counts.length, 16), div(ids.length, 8))) {
+                // InvalidReportData
+                mstore(0x00, 0xc6726884)
+                revert(0x1c, 0x04)
+            }
+
+            // counts.length % 16 != 0
+            if iszero(iszero(mod(counts.length, 16))) {
+                // InvalidReportData
+                mstore(0x00, 0xc6726884)
+                revert(0x1c, 0x04)
+            }
 
-    function validate(bytes calldata ids, bytes calldata counts) internal pure {
-        if (
-            counts.length / 16 != ids.length / 8 ||
-            ids.length % 8 != 0 ||
-            counts.length % 16 != 0
-        ) {
-            revert InvalidReportData();
+            // ids.length % 8 != 0
+            if iszero(iszero(mod(ids.length, 8))) {
+                // InvalidReportData
+                mstore(0x00, 0xc6726884)
+                revert(0x1c, 0x04)
+            }
+
+            count := div(ids.length, 8)
```

Comment:
it adds few bytes and some complexity, so i'd suggest to left it as is
```solidity
        if (
            counts.length / 16 != ids.length / 8 ||
            ids.length % 8 != 0 ||
            counts.length % 16 != 0
        ) {
            revert InvalidReportData();
        }
        count = ids.length / 8;
```
free size 838 > 789

Reply:
:(